Refactor layerChooser, reorganize boot.js, refactor android-specific code

johnd0e opened the pull request:
* rename android.->app., to be more general (as we have iOS app as well)

* collect (most of) android-specific code into single file `app.js`
  - No more numerous `if (typeof android !== 'undefined' && android && android...` conditions
    in random places of code - all relevant stuff is moved to android.js (whenever possible).
    In those places where still needed - new property `window.isApp` is used.
  - Instead of conditions - some functions is directly overwritten (when needed).

* layerChooser: get scattered code together into separate class.
  - android-specific methods included only when needed (android.js)
  - new method toggleLayer, now used also in IITCm (instead of old incomplete java implementation)
    + so longlick in layerchooser now acts consistently both on desktop/mobile
    + IITCm: fix exception when longclick not on layer.
  - new plugins:
    + hide-portal-levels
    + remove-extra-layers

* boot.js: 
  - sidebar stuff is separated in `sidebar.js`
  - map stuff is separated in `map.js`
  - map options are now defined in `window.mapOptions`
    (overridable by plugins with priority `'boot'`)


<!--
This is the same as #483 but in distinct commits, for easier review.
([more to come](https://github.com/johnd0e/ingress-intel-total-conversion/compare/refactor-map..refactor-layerchooser))
-->

Commit:
boot.js: separate map-related stuff into separate module 'map.js'

Commit:
boot.js: replace window.extendLeaflet() with setupIngressMarkers()/setupCRS()

Commit:
map.js: add content of location.js

Commit:
map.js: do not expose getPosition and storeMapPosition

Commit:
map.js: simplify map position storing func

Commit:
map.js: refactor getPosition()

Commit:
map.js: refactor default overlays creation

Commit:
map.js: separate code into createDefaultOverlays() func

Commit:
map.js: do not expose setMapBaseLayer()

Commit:
map.js: map options are now defined in window.mapOptions

(overridable by plugins with priority 'boot')

Commit:
layerchooser.js: put together all related code from boot.js/utils_misc.js

utils_misc.js: window.layerGroupLength was unused

Commit:
layerchooser.js: do not expose setupLayerChooserSelectOne() and setupLayerChooserStatusRecorder()

Commit:
layerchooser.js: setup own class and expose it as `window.LayerChooser`

..instead of monkey-patching window.layerChooser instance

Android: as _update is always hooked - no need to force call getLayers() in setupLayerChooserApi

Commit:
layerchooser.js: separate android-related handling

Remove try/catch as debounced function is safe to use:
even if JSON.stringify() throws - that does not stop main thread

Mark `getLayers` method as deprecated as it does not have much general-use application

Commit:
layerchooser.js: refactor setupLayerChooserSelectOne() and include into LayerChooser

New private method _toggleOverlay() common both to desktop and Android app
(instead of it's own incomplete reimplementation),
so longclick in layerchooser now acts consistently both on desktop/mobile.

_toggleOverlay() ignores 'DEBUG Data Tiles' and faction layers.

Android: fix exception when longclick not on layer.

Commit:
layerchooser.js: refactor showLayer() in order to make it useful public method

Now accepts layer by it's name too

Commit:
layerchooser.js: put overlays status tracking into the class

..so no need to setupLayerChooserStatusRecorder() separately

window.setupLayerChooserApi is not exposed anymore,
the rest of the code inlined to map.js

Commit:
utils_misc.js: add Array/find polyfill

..and update some others

Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find#polyfill

Commit:
layerchooser.js: handle overlays 'persistency' automatically

Commit:
layerchooser.js: make overlayStatus and updateDisplayedLayerGroup() private

Commit:
layerchooser.js: deprecate window.isLayerGroupDisplayed

As it is not straight-forward to use
(may be confusing returning `true` when overlay is not even in layerChooser).
For most cases `map.hasLayer` should be preferred.

The logic moved into LayerChooser class.

Commit:
layerchooser.js: new method getLayerByName(), and `removeLayer` now accepts name as well

Commit:
layerchooser.js: track/store baselayer changes

Commit:
layerchooser.js: deprecate window.addLayerGroup in favor of layerChooser.addOverlay

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -72,21 +36,58 @@ var LayerChooser = L.Control.Layers.extend({
+    // hook some additional code into the LayerControl so it's easy for the mobile app to interface with it
+    _setAndroidLayers: debounce(function () {
+      var l = this.getLayers();
+      android.setLayers((l.baseLayers), JSON.stringify(l.overlayLayers));
```
JSON missing

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -1,5 +1,26 @@
+
+  removeLayer: function (layer) {
+    if (layer && layer._statusTracking) {
+      layer.off('add remove', this._statusTracking, this);
```
`layer._statusTracking` ?

le-jeu commented on plugins/draw-tools.js:
```diff
@@ -713,10 +713,10 @@ window.plugin.drawTools.initMPE = function(){
       console.log('DRAWTOOLS: reset all drawn items (func_post)');
 
-      if (window.plugin.crossLinks !== undefined && window.overlayStatus['Cross Links']) {
+      if (window.plugin.crossLinks) {
```
I assume crosslink and destroyed plugins have to do this on their own, tracking whether their respective layer are displayed

johnd0e replied:
They should listen for draw-tools 'load' hook. But that is out of scope of this PR.

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -33,6 +34,11 @@ var LayerChooser = L.Control.Layers.extend({
+      layer._statusTracking = function () {
+        localStorage['iitc-base-map'] = name;
+      };
+      layer.on('add', layer._statusTracking);
```
+`,this)`

johnd0e replied:
Actually not

le-jeu replied:
right

Commit:
layerchooser.js: deprecate removeLayerGroup in favor of layerChooser.removeLayer

layerChooser.removeLayer: optional `keepOnMap` argument prevents layer from removing from map.

Commit:
layerchooser.js: do not track status of layer by setting it's `notPersistent` option.

Commit:
layerchooser.js: fire 'longclick' event on overlay

..including ability to override default (_toggleOverlay) action
with e.preventDefault() method.

Note: that may look redundant, but it seems that there is not better way to detect
if default action is still needed after all events processed.
E.g. consider such case:

mylayer.on('longclick', function (e) {
  if (!e.originalEvent.metaKey) { return; }
  e.preventDefault();
  doThings();
});

// then in different plugin
mylayer.on('longclick', function (e) {
  if (e.type !== 'taphold') { return; }
  e.preventDefault();
  doOtherThings();
});

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -6,6 +6,7 @@
  * Provides 'persistence' of layers display state between sessions, saving it to localStorage.
  * Every overlay is added to map automatically if it's last state was active.
  * When no record exists - active is assumed, except when layer has option `defaultDisabled`.
+ * And with `notPersistent` layer option it is possible to suppress it's status tracking.
```
Nice

Commit:
layerchooser.js: possibility to enforce specific layers order

..specifying `sortPriority` in layer's option

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -12,16 +12,39 @@
   _addLayer: function (layer, name, overlay) {
     L.Control.Layers.prototype._addLayer.apply(this, arguments);
+    // provide stable sort order
+    if (!layer.options.sortPriority) {
```
technically, `0` may be a correct value

johnd0e replied:
`if (!('sortPriority' in layer.options)) {`

le-jeu replied:
or `!== undefined`
I use this one, but yours is more expressive :+1:

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -133,6 +133,47 @@ var LayerChooser = L.Control.Layers.extend({
+
+    // @event overlaylongclick: LayersControlEventExt
+    // Fired on map
+    this._map.fire('overlaylongclick', L.extend(obj, el));
```
`L.extend(obj, el)` is this common pattern in Leaflet ?

johnd0e replied:
Yes.
Do you have some concerns regarding our case?

le-jeu replied:
no, it looks weird to extend event data with a layer
I expect `L.extend(obj.target/progabateTarget/sourceTarget, el)` instead
following the https://leafletjs.com/reference-1.7.1.html#event description

johnd0e replied:
`target` is set by leaflet: https://github.com/Leaflet/Leaflet/blob/0f904a515879fcd08f69b7f51799ee7f18f23fd8/src/core/Events.js#L179

> no, it looks weird to extend event data with a layer

`el` is not layer, and we do basically the same as `L.Control.Layers`: https://github.com/Leaflet/Leaflet/blob/0f904a515879fcd08f69b7f51799ee7f18f23fd8/src/control/Control.Layers.js#L303

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -133,6 +133,47 @@ var LayerChooser = L.Control.Layers.extend({
+    // @event overlaylongclick: LayersControlEventExt
+    // Fired on map
+    this._map.fire('overlaylongclick', L.extend(obj, el));
+    if (!defaultPrevented) {
```
why not, but I don't see use case since the default is still toggleOverlay
do you expect an override like:
```js
map.on('overlaylongclick', function (e) { 
  e.preventDefault(); 
  doThings();
});

johnd0e replied:
I expect something like this:
```js
map.on('overlaylongclick', function (e) {
  if (!e.originalEvent.metaKey) { return; }
  e.preventDefault();
  doThings();
});
// then in different plugin
map.on('overlaylongclick', function (e) {
  if (!e.originalEvent.ctrlKey) { return; }
  e.preventDefault();
  doOtherThings();
});
```
So you are right, explicit `e.preventDefault();` is not convenient, but I have not found better way to detect if default action is still needed after all events processed.

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -253,13 +257,7 @@ window.addLayerGroup = function (name, layerGroup, defaultDisplay) {
   window.layerChooser.addOverlay(layerGroup, name);
 };
 
+// !!deprecated: use `layerChooser.removeLayer` directly (https://leafletjs.com/reference.html#control-layers-removelayer)
```
leaflet doc is confusing since `layerChooser.removeLayer` removes the layer from the map by default

johnd0e replied:
But a line below you can see clear usage sample that clarifies any confusion)

Commit:
Sample plugin: hide-portal-levels

Key points:
- it hides portal levels, replacing them with single entry
- hidden layers enforce to be active
- toggle by longclicking on portal levels

Commit:
map.js: put all location-related in one place

Commit:
Sample plugin: collapsible-layers

Key points:
- it keeps original levels state (indicating inactive layers by label style)
- it toggles by longclicking on faction layers (by default), that may be not so intuitive,
  but has some benefit: it plays good together with standard toggling feature
- it is implemented in reusable way, as leaflet class (as sample it also toggles overlays)

Commit:
layerchooser.js: read overlays status from localstorage on initialize only

Commit:
layerchooser.js: restore window.overlayStatus (for compatibility)

A bit smarter compability hook implementation

Commit:
Sample plugin: remove-extra-layers

Commit:
layerchooser.js: separate layer's name from label in layerChooser

Name must be consistent, label can be changed (styled with html etc)
(for IITCm tags are stripped for now)

New public method `setLabel()`.

le-jeu commented on plugins/collapsible-layers.js:
```diff
@@ -56,22 +63,30 @@ function setup () {
+          disabled.all = true;
+        }
+
+        var label;
```
```suggestion
        var label = this._name;
```

le-jeu commented on plugins/collapsible-layers.js:
```diff
@@ -56,22 +63,30 @@ function setup () {
         }
-        this._ctrl.addOverlay(this, name);
+        this.options.defaultDisabled = disabled.all;
+        this._ctrl.addOverlay(this, label ||  this._name);
```
```suggestion
        this._ctrl.addOverlay(this, label);
```

johnd0e replied:
I wrote exactly like you suggested first, but then changed for the sake of expression,

le-jeu replied:
in this case, I would separate calls with the `if` above (style anyway)

le-jeu commented on plugins/collapsible-layers.js:
```diff
@@ -14,26 +14,33 @@ collapsibleLayers.initial = true; // collapsed
+
+      this.on('add remove', function (e) { // store status
+        this.eachLayer(function (layer) {
+          this._storeOverlayState(layer._name, e.type === 'add'); //??
```
can this be done by adding/removing the layers from the map ?

johnd0e replied:
Unfortunately not, as layers are not tracked anymore after they removed from layerChooser.

But I do not like using private methods as well, so suggestions are welcome.

johnd0e replied:
Atm I'm considering another approach: additional 'active' argument to addOverlay/BaseLayer (to enforce desirable state despite of what is recorded in last status), and use that on layers re-adding.

But it will not help to store states while layers are collapsed (e.g.: set 'Portals' to unchecked, then reload - and they are checked again).
Pretty minor case, but I'm still not completely sure.

le-jeu replied:
what about not storing anything ? 
the point of the layer is to enable/disable all collapsed layer in one click. 
It's fair not to be persistent changes

johnd0e replied:
Consider such case:
- collapse all-enabled levels into single entry
- disable that entry
- expand

What state do you expect?
Without storing state - all be (unexpectedly) enabled.

le-jeu replied:
this can be handled line 93, calling `showLayer(layer, false)` for disabled layer

johnd0e replied:
That is far from optimal: thousands of layers will be added and then immediately removed.

johnd0e replied:
Experimental: https://github.com/IITC-CE/ingress-intel-total-conversion/pull/486/commits/a32772057504212b743b6721734e773c43460e7e

le-jeu commented on plugins/collapsible-layers.js:
```diff
@@ -14,26 +14,33 @@ collapsibleLayers.initial = true; // collapsed
+
       this._name = options.name || match;
+      this._ctrl = window.layerChooser;
       this._layers = {};
```
`this._layers` is never used

johnd0e replied:
It is used in parent class.
There is also `clearLayers` method but it has undesirable (for our case) side effects

le-jeu replied:
then, calling `LayerGroup.initialize.call(this.[],options)` sounds more idiomatic

johnd0e replied:
Still heavily dependent on parent implementation, so in the end there is no difference.

Commit:
layerchooser.js: .addOverlay process option 'enable'

..to enforce some state regardless of previously recorded

Commit:
layerchooser.js: addBaseLayer/addOverlay options .persistent/.default/.sortPriority options

..instead of using layer's options (.notPersistent/.defaultDisabled/...)

Commit:
layerchooser.js: stop polluting of layer object with numerous layerChooser's private properties

Use single `_chooser` property.

Commit:
layerchooser.js: use option object in removeLayer (instead of boolean keepOnMap arg)

Commit:
layerchooser.js: use _layers to store extra data

It is also in layer._chooser - in order to keep data in case of removal/re-adding,
but not used for any other purpose.

Make `layerInfo` public method.

Rename `el` and `info` occurences to `data` - for consistence.

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -238,29 +275,25 @@ var LayerChooser = L.Control.Layers.extend({
-    label = label || info.layer._chooser.name;
-    info.name = label;
-    var nameEl = this._layerControlInputs[idx].closest('label').querySelector('span');
+    label = label;
```
`label = label || data.name` ?

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -371,11 +404,11 @@ var LayerChooser = L.Control.Layers.extend({
         layerId: idx,
-        name: this._stripHtmlTags(info.name), // IITCm does not support html in layers labels
-        active: this._map.hasLayer(info.layer)
+        name: this._stripHtmlTags(data.label), // IITCm does not support html in layers labels
```
+ `|| data.name`

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -99,6 +122,17 @@ var LayerChooser = L.Control.Layers.extend({
+      name: obj.label || obj.name
+    });
+    obj.labelEl = labelEl;
+    obj.inputEl = this._layerControlInputs[this._layerControlInputs.length-1];
```
`inputEL` is not used

johnd0e replied:
But can be used in plugins

le-jeu commented on core/code/layerchooser.js:
```diff
@@ -238,29 +275,25 @@ var LayerChooser = L.Control.Layers.extend({
-    var nameEl = this._layerControlInputs[idx].closest('label').querySelector('span');
+    label = label;
+    data.label = label;
+    var nameEl = data.labelEl.closest('label').querySelector('span');
```
to my understanding, `L.LayerChooser._addItem` returns the `<label>` already

johnd0e replied:
Right

Commit:
layerchooser.js: include layer data into 'longclick' event (instead of internal idx)

.layerInfo and .getLayer accept labelEl (HTMLElement) as well

Commit:
android.js: incapsulate all app-specific code

Skeleton for now.

New property `window.isAndroid`

Commit:
android.js: move code from smartphone.js

Commit:
android.js: move code from dialog.js

Commit:
android.js: move code from map.js

Commit:
android.js: move code from panes.j

Commit:
android.js: move code from layerchooser.js

Commit:
android.js: move androidCopy implementation from utils_misc.js

Commit:
android.js: move saveFile implementation from utils_file.js

Commit:
Refactor 'fixed addons' setup

Get rid of deprecated setPermaLink and androidPermalink

Commit:
Refactor #linkdetails rendering

New private function window.renderPortalUrl

Commit:
utils_misc.js: refactor showPortalPosLinks

johnd0e commented:
New batch of commits
@le-jeu

le-jeu commented on core/code/map_data_request.js:
```diff
@@ -60,7 +60,7 @@ window.MapDataRequest = function() {
 
   // delay before repeating the render loop. this gives a better chance for user interaction
-  this.RENDER_PAUSE = (typeof android === 'undefined') ? 0.1 : 0.2; //100ms desktop, 200ms mobile
+  this.RENDER_PAUSE = window.isAndroid ? 0.1 : 0.2; // 100ms desktop, 200ms mobile
```
```suggestion
  this.RENDER_PAUSE = !window.isAndroid ? 0.1 : 0.2; // 100ms desktop, 200ms mobile
```

le-jeu commented on core/code/map.js:
```diff
@@ -206,10 +206,6 @@ window.setupMap = function() {
-  // hide layer chooser if booted with the iitcm android app
-  if (window.isAndroid && android.setLayers) {
-    $('.leaflet-control-layers').hide();
-  }
```
where is this moved ?

johnd0e replied:
to `android.js`

le-jeu commented on core/code/utils_misc.js:
```diff
@@ -106,6 +106,31 @@ window.aboutIITC = function () {
+
+  window.artifact.setup();
+
+  window.RegionScoreboardSetup();
```
```suggestion
  window.RegionScoreboard.setup();
```

le-jeu commented on core/code/utils_misc.js:
```diff
@@ -106,6 +106,31 @@ window.aboutIITC = function () {
+    })
+    .on({
+      mouseover: setPermaLink,
+      click: setPermaLink
```
`setPermalink` is removed

le-jeu commented on core/code/android.js:
```diff
@@ -112,6 +112,18 @@ window.runOnAndroidBeforeBoot = function () {
+      });
+      $('linkdetails').append($('<aside>').append(shareLink));
+    };
+  }
```
this pattern (creating a HTMLElement, and fill if afterward) looks wrong to me in this context.
What about hooking on `renderDetailUpdated` instead ? Or make the function return a `<div>` directly ?

johnd0e replied:
> What about hooking on renderDetailUpdated instead?

Thwt would be perfect, but there is a condition: https://github.com/IITC-CE/ingress-intel-total-conversion/blob/4d46ab7178bc2c07252c05758a393ac34b037b2d/core/code/portal_detail_display.js#L155-L157

> Or make the function return a `<div>` directly ?

Well, there are 2 elements.. Put them under one div?

le-jeu replied:
> Thwt would be perfect, but there is a condition: 

yes, the hook doesn't match so that won't work

> Well, there are 2 elements.. Put them under one div? 

there is only one `<div class="linkDetails"/>` so that matches

note: Raw portal also inject it's "Raw data" button in this div

(anyway, that's only a pattern I don't commonly use. I prefer function that returns element smth that function that display smth when the choice)

le-jeu commented on core/code/portal_detail_display.js:
```diff
@@ -9,6 +9,24 @@ window.resetScrollOnNewPortal = function() {
 
+// to be ovewritten in android.js
+window.renderPortalUrl = function (lat, lng, title, guid) {
+  var linkDetails = $('linkdetails');
```
isn't it `'.linkdetails'` ?

Commit:
sidebar.js: move code from boot.js

Commit:
sidebar.js: move code from utils_misc.js

Commit:
sidebar.js: do not expose private functions

Commit:
region_scoreboard.js: do not expose RegionScoreboard obj

Commit:
Minor code rearrangements in boot.js, map.js, sidebar.js

Commit:
lint a bit

johnd0e commented:
@le-jeu 
Final version

johnd0e commented:
May be better change names
`android.js -> app.js`
`isAndroid() -> isApp`

le-jeu commented on core/code/sidebar.js:
```diff
@@ -0,0 +1,264 @@
+  + '<div>'
+  + '  <a href="'+'@url_homepage@'+'" target="_blank">IITC Homepage</a> |' +
+    '  <a href="'+'@url_tg@'+'" target="_blank">Telegram channel</a><br />'
+  + '   On the script�s homepage you can:'
```
the `’` is not unicode here

le-jeu commented:
> May be better change names `android.js -> app.js` `isAndroid() -> isApp`

assuming the iOS app implement parts of the android object, it sounds good wrt current PR state

le-jeu commented on core/code/sidebar.js:
```diff
@@ -0,0 +1,264 @@
+    .click(aboutIITC)
+    .appendTo('#toolbox');
+
+  window.artifact.setup();
```
while this add a button in the toolbox, I'd rather put it back into the main boot sequence, as it's a whole intel feature

same could be said about regionscoreboard
that makes me wonder if a some `~IITC.sidebar.addToolboxLink` (to be renamed) would be a better API

johnd0e replied:
It's all about intel features, imho. Compare with `missions`.
So.. I'm not sure.

For me it belongs to sidebar because we cannot see it otherwise - only by clicking sidebar links.

modos189 commented on core/code/sidebar.js:
```diff
@@ -0,0 +1,264 @@
+window.setupSidebar = function() {
```
Something is wrong with file encoding

modos189 commented on core/code/sidebar.js:
```diff
@@ -0,0 +1,264 @@
+      '#sidebar { width:'+(SIDEBAR_WIDTH + HIDDEN_SCROLLBAR_ASSUMED_WIDTH + 1 /*border*/)+'px;  } ',
+      '#sidebartoggle { right:'+(SIDEBAR_WIDTH+1)+'px;  } ',
+      '#scrollwrapper  { width:'+(SIDEBAR_WIDTH + 2*HIDDEN_SCROLLBAR_ASSUMED_WIDTH)+'px; right:-'+(2*HIDDEN_SCROLLBAR_ASSUMED_WIDTH-2)+'px } ',
+      '#sidebar > * { width:'+(SIDEBAR_WIDTH+1)+'px;  }'].join('\n')
```
These styles have a higher priority than the styles in smartphone.css.

For example, on mobile the portal information window is not stretched across the entire width

modos189 commented on core/code/android.js:
```diff
@@ -0,0 +1,168 @@
+    // hook some additional code into the LayerControl so it's easy for the mobile app to interface with it
+    window.LayerChooser.include({
+      _setAndroidLayers: debounce(function () {
+        var l = this.getLayers();
```
I see that it works, but how? Is there a Leaflet method getLayers() that returns the data in the form we want?

johnd0e replied:
We have own `.getLayers()`, see `layerchooser.js`

modos189 commented:
> May be better change names `android.js -> app.js` `isAndroid() -> isApp`

Then it would be a good to make alias `window.android` → `window.app`

johnd0e commented:
@le-jeu

> assuming the iOS app implement parts of the android object, it sounds good wrt current PR state

So you do support renaming?

johnd0e commented:
@le-jeu 
There are also 3 demo plugins (in Layers category).
Should we keep them? And which of them?

Commit:
Rename android->app

Commit:
android.js->app.js

le-jeu commented:
> @le-jeu There are also 3 demo plugins (in Layers category). Should we keep them? And which of them?

I forgot about the demo plugins
They could end into the Wiki. The huge one is somewhat a combination of the small ones.
Maybe keep only `hide-portal-level` or `collapsible-layer`

johnd0e commented:
@McBen @MysticJay

Commit:
Exclude demo plugin collapsible-layers
